File: WMF Framework/NSManagedObjectContext+WMFKeyValue.m
#import <WMF/NSManagedObjectContext+WMFKeyValue.h>
#import <WMF/WMFKeyValue+CoreDataProperties.h>
#import <WMF/WMFLogging.h>

@implementation NSManagedObjectContext (WMFKeyValue)

- (nullable NSArray<WMFKeyValue *> *)wmf_keyValuesForKey:(NSString *)key fetchLimit:(NSInteger)fetchLimit {
    NSFetchRequest *request = [WMFKeyValue fetchRequest];
    request.predicate = [NSPredicate predicateWithFormat:@"key == %@", key];
    if (fetchLimit > 0) {
        request.fetchLimit = fetchLimit;
    }
    NSError *keyValueFetchError = nil;
    NSArray<WMFKeyValue *> *results = [self executeFetchRequest:request error:&keyValueFetchError];
    if (keyValueFetchError) {
        DDLogError(@"Error fetching key value: %@", keyValueFetchError);
    }
    return results;
}
- (nullable WMFKeyValue *)wmf_keyValueForKey:(NSString *)key {
    NSArray<WMFKeyValue *> *results = [self wmf_keyValuesForKey:key fetchLimit:1];
    return results.firstObject;
}

- (nullable id)wmf_valueOfClass:(Class) class forKey:(NSString *)key {
    WMFKeyValue *keyValue = [self wmf_keyValueForKey:key];
    id value = keyValue.value;
    if ([value isKindOfClass:class]) {
        return value;
    } else {
        return nil;
    }
}

    - (nullable NSNumber *)wmf_numberValueForKey : (NSString *)key {
    return [self wmf_valueOfClass:[NSNumber class] forKey:key];
}

- (nullable NSString *)wmf_stringValueForKey:(NSString *)key {
    return [self wmf_valueOfClass:[NSString class] forKey:key];
}

- (nullable NSArray *)wmf_arrayValueForKey:(NSString *)key {
    return [self wmf_valueOfClass:[NSArray class] forKey:key];
}

- (WMFKeyValue *)wmf_setValue:(nullable id<NSCoding>)value forKey:(NSString *)key {
    NSArray<WMFKeyValue *> *results = [self wmf_keyValuesForKey:key fetchLimit:0];
    if (results.count > 1) {
        // failsafe to delete extra key value objects
        NSArray<WMFKeyValue *> *subarray = [results subarrayWithRange:NSMakeRange(1, results.count - 1)];
        for (WMFKeyValue *value in subarray) {
            [self deleteObject:value];
        }
    }
    WMFKeyValue *keyValue = results.firstObject;
    if (!keyValue) {
        keyValue = [NSEntityDescription insertNewObjectForEntityForName:@"WMFKeyValue" inManagedObjectContext:self];
        keyValue.key = key;
    }
    keyValue.value = value;
    return keyValue;
}

@end

File: WMF Framework/NSManagedObjectContext+WMFKeyValue.swift
import CoreData
import CocoaLumberjackSwift

/// Key-value storage on a managed object context, backed by the WMFKeyValue entity.
/// Converted from Objective-C so managed-object access here is visible to the Swift
/// strict-concurrency checker. All methods must be called on the context's queue
/// (inside perform/performAndWait or on the context's thread), same as before.
extension NSManagedObjectContext {

    private func wmf_keyValues(forKey key: String, fetchLimit: Int) -> [WMFKeyValue] {
        let request = NSFetchRequest<WMFKeyValue>(entityName: "WMFKeyValue")
        request.predicate = NSPredicate(format: "key == %@", key)
        if fetchLimit > 0 {
            request.fetchLimit = fetchLimit
        }
        do {
            return try fetch(request)
        } catch {
            DDLogError("Error fetching key value: \(error)")
            return []
        }
    }

    @objc(wmf_keyValueForKey:)
    public func wmf_keyValue(forKey key: String) -> WMFKeyValue? {
        return wmf_keyValues(forKey: key, fetchLimit: 1).first
    }

    private func wmf_value<T>(of type: T.Type, forKey key: String) -> T? {
        return wmf_keyValue(forKey: key)?.value as? T
    }

    @objc(wmf_numberValueForKey:)
    public func wmf_numberValue(forKey key: String) -> NSNumber? {
        return wmf_value(of: NSNumber.self, forKey: key)
    }

    @objc(wmf_stringValueForKey:)
    public func wmf_stringValue(forKey key: String) -> String? {
        return wmf_value(of: NSString.self, forKey: key) as String?
    }

    @objc(wmf_arrayValueForKey:)
    public func wmf_arrayValue(forKey key: String) -> NSArray? {
        return wmf_value(of: NSArray.self, forKey: key)
    }

    @discardableResult
    @objc(wmf_setValue:forKey:)
    public func wmf_setValue(_ value: NSCoding?, forKey key: String) -> WMFKeyValue {
        let results = wmf_keyValues(forKey: key, fetchLimit: 0)
        if results.count > 1 {
            // failsafe to delete extra key value objects
            for extraValue in results.dropFirst() {
                delete(extraValue)
            }
        }
        let keyValue: WMFKeyValue
        if let existing = results.first {
            keyValue = existing
        } else {
            keyValue = NSEntityDescription.insertNewObject(forEntityName: "WMFKeyValue", into: self) as! WMFKeyValue
            keyValue.key = key
        }
        keyValue.value = value
        return keyValue
    }
}
